extensions/data-plane/data-plane-google-bigquery/src/main/java/org/eclipse/edc/connector/dataplane/gcp/bigquery/pipeline/BigQueryDataSource.java

```diff
@@ -0,0 +1,300 @@
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.PipedInputStream;
+import java.io.PipedOutputStream;
+import java.util.Map;
+import java.util.Objects;
+import java.util.UUID;
+import java.util.concurrent.ExecutorService;
+import java.util.stream.Stream;
+
+import static org.eclipse.edc.connector.dataplane.spi.pipeline.StreamResult.error;
+import static org.eclipse.edc.connector.dataplane.spi.pipeline.StreamResult.success;
+import static org.eclipse.edc.spi.constants.CoreConstants.EDC_NAMESPACE;
+
+/**
+ * Reads data using the BigQuery API (REST) and transfer it in JSON format.
+ */
+public class BigQueryDataSource implements DataSource {
+    /**
+     * Maximum number of rows per page retrieved by the query.
+     */
+    private static final int BIGQUERY_PAGE_SIZE = 4;
+    private BigQueryRequestParams params;
+    private String requestId;
+    private Monitor monitor;
+    private BigQueryConfiguration configuration;
+    private BigQueryTarget target;
+    private ExecutorService executorService;
+    private BigQuery bigQuery;
+    private GoogleCredentials credentials;
+
+    private BigQueryDataSource() {
+    }
+
+    @Override
+    public void close() {
+    }
+
+    @Override
+    public StreamResult<Stream<Part>> openPartStream() {
+        try {
+            var query = params.getQuery();
+            var sinkAddress = params.getSinkAddress();
+            var queryConfigBuilder = QueryJobConfiguration.newBuilder(query);
+            var destinationTable = params.getDestinationTable();
+            var customerName = sinkAddress.getStringProperty(BigQueryServiceSchema.CUSTOMER_NAME);
+            if (customerName != null) {
+                queryConfigBuilder.setLabels(Map.of("customer", customerName));
+            }
+
+            setNamedParameters(queryConfigBuilder, sinkAddress);
+
+            // Use standard SQL syntax for queries.
+            // See: https://cloud.google.com/bigquery/sql-reference/
+            queryConfigBuilder = queryConfigBuilder
+                .setUseLegacySql(false)
+                .setUseQueryCache(true);
+            if (destinationTable != null) {
+                queryConfigBuilder = queryConfigBuilder.setDestinationTable(target.getTableId(destinationTable));
+            }
+
+            var queryConfig = queryConfigBuilder.build();
+            var jobId = JobId.of(UUID.randomUUID().toString());
+            var jobInfo = JobInfo.newBuilder(queryConfig).setJobId(jobId).build();
+            var queryJob = bigQuery.create(jobInfo);
+
+            // The output stream object is passed to the thread lambda, do not use try with resources.
+            var outputStream = new PipedOutputStream();
+            // The input stream object is passed to the sink thread, do not use try with resources.
+            var inputStream = new PipedInputStream(outputStream);
+            var part = new BigQueryPart("allRows", inputStream);
+
+            executorService.submit(() -> {
```

Comment:
nit: can be replaced with an expression lambda

Reply:
Done